src/app/payments/helpers.ts

```diff
@@ -0,0 +1,296 @@
+    },
+    async () => {
+      const midPriceRatio = await getMidPriceRatio()
+      if (midPriceRatio instanceof Error) return midPriceRatio
+
+      const btcPaymentAmount = {
+        amount: BigInt(Math.ceil(Number(amount.amount) / midPriceRatio)),
+        currency: WalletCurrency.Btc,
+      }
+
+      addAttributesToCurrentSpan({
+        "btcFromUsdMidPriceFn.midPriceRatio": midPriceRatio,
+        "btcFromUsdMidPriceFn.incoming.amount": Number(amount.amount),
+        "btcFromUsdMidPriceFn.incoming.unit":
+          amount.currency === WalletCurrency.Usd
+            ? ExchangeCurrencyUnit.Usd
+            : ExchangeCurrencyUnit.Btc,
+        "btcFromUsdMidPriceFn.outgoing.amount": Number(btcPaymentAmount.amount),
+        "btcFromUsdMidPriceFn.outgoing.unit":
+          btcPaymentAmount.currency === WalletCurrency.Btc
+            ? ExchangeCurrencyUnit.Btc
+            : ExchangeCurrencyUnit.Usd,
+      })
+
+      return btcPaymentAmount
+    },
+  )
+
+export const getCurrentPriceInCentsPerSat = async (): Promise<
+  CentsPerSatsRatio | PriceServiceError
+> => {
+  const price = await getCurrentPrice()
+  if (price instanceof Error) return price
+
+  return (price * CENTS_PER_USD) as CentsPerSatsRatio
+}
+
+export const getMidPriceRatio = async (): Promise<
+  CentsPerSatsRatio | DealerPriceServiceError | PriceServiceError
+> => {
+  let midPriceRatio = usdHedgeEnabled
+    ? await dealer.getCentsPerSatsExchangeMidRate()
+    : await getCurrentPriceInCentsPerSat()
+  if (midPriceRatio instanceof Error && usdHedgeEnabled) {
+    recordExceptionInCurrentSpan({
+      error: midPriceRatio,
+      level: ErrorLevel.Critical,
+    })
+    midPriceRatio = await getCurrentPriceInCentsPerSat()
+  }
+
+  return midPriceRatio
+}
+
+export const constructPaymentFlowBuilder = async ({
+  senderWallet,
+  invoice,
+  uncheckedAmount,
+  usdFromBtc,
+  btcFromUsd,
+}: {
+  senderWallet: Wallet
+  invoice: LnInvoice
+  uncheckedAmount?: number
+  usdFromBtc: (amount: BtcPaymentAmount) => Promise<UsdPaymentAmount | ApplicationError>
+  btcFromUsd: (amount: UsdPaymentAmount) => Promise<BtcPaymentAmount | ApplicationError>
+}): Promise<LPFBWithConversion<WalletCurrency, WalletCurrency> | ApplicationError> => {
+  const lndService = LndService()
+  if (lndService instanceof Error) return lndService
+  const paymentBuilder = LightningPaymentFlowBuilder({
+    localNodeIds: lndService.listAllPubkeys(),
+    usdFromBtcMidPriceFn,
+    btcFromUsdMidPriceFn,
+  })
+  const builderWithInvoice = uncheckedAmount
+    ? paymentBuilder.withNoAmountInvoice({ invoice, uncheckedAmount })
+    : paymentBuilder.withInvoice(invoice)
+
+  const builderWithSenderWallet = builderWithInvoice.withSenderWallet(senderWallet)
+
+  let builderAfterRecipientStep:
+    | LPFBWithRecipientWallet<WalletCurrency, WalletCurrency>
+    | LPFBWithError
+  if (
+    !(builderWithSenderWallet.isIntraLedger() instanceof Error) &&
```

Comment:
Confused as to why this function can return an Error...

Reply:
You're right, this needed to be cleaned up. There is a code path where errors come through, but it's irrelevant because the rest of the logic is resolved via an `LPFBWithError` object.

Cleaned up: https://github.com/GaloyMoney/galoy/pull/1237/commits/ed23db6d17a90c1fc562c19815597c54ffa7e84e